pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,654 @@
+		driver := &s3Driver{
+			Creds:    creds,
+			CopyFrom: o.AttemptS3BucketCopy,
+		}
+		url := ref.DockerClientDefaults().RegistryURL()
+		return driver.Repository(ctx, url, ref.RepositoryName(), o.Insecure)
+	default:
+		return nil, fmt.Errorf("unrecognized destination type %s", t)
+	}
+}
+
+// includeDescriptor returns true if the provided manifest should be included.
+func (o *pushOptions) includeDescriptor(d *manifestlist.ManifestDescriptor) bool {
+	if o.OSFilter == nil {
+		return true
+	}
+	return o.OSFilter.MatchString(fmt.Sprintf("%s/%s", d.Platform.OS, d.Platform.Architecture))
+}
+
+// ErrAlreadyExists may be returned by the blob Create function to indicate that the blob already exists.
+var ErrAlreadyExists = fmt.Errorf("blob already exists in the target location")
+
+var schema2ManifestOnly = distribution.WithManifestMediaTypes([]string{
+	manifestlist.MediaTypeManifestList,
+	schema2.MediaTypeManifest,
+})
+
+func (o *pushOptions) Run() error {
+	tree := buildTargetTree(o.Mappings)
+
+	creds := importer.NewLocalCredentials()
+	ctx := apirequest.NewContext()
+
+	rt, err := rest.HTTPWrappersForConfig(&rest.Config{}, http.DefaultTransport)
+	if err != nil {
+		return err
+	}
+	insecureRT, err := rest.HTTPWrappersForConfig(&rest.Config{TLSClientConfig: rest.TLSClientConfig{Insecure: true}}, http.DefaultTransport)
+	if err != nil {
+		return err
+	}
+	srcClient := importer.NewContext(rt, insecureRT).WithCredentials(creds)
+	toContext := importer.NewContext(rt, insecureRT).WithActions("pull", "push")
+
+	var errs []error
+	for _, src := range tree {
+		srcRepo, err := srcClient.Repository(ctx, src.ref.DockerClientDefaults().RegistryURL(), src.ref.RepositoryName(), o.Insecure)
+		if err != nil {
+			errs = append(errs, retrieverError{err: fmt.Errorf("unable to connect to %s: %v", src.ref, err), src: src.ref})
+			continue
+		}
+
+		manifests, err := srcRepo.Manifests(ctx)
+		if err != nil {
+			errs = append(errs, retrieverError{src: src.ref, err: fmt.Errorf("unable to access source image %s manifests: %v", src.ref, err)})
+			continue
+		}
+
+		var tagErrs []retrieverError
+		var digestErrs []retrieverError
+
+		// convert source tags to digests
+		for srcTag, pushTargets := range src.tags {
+			desc, err := srcRepo.Tags(ctx).Get(ctx, srcTag)
+			if err != nil {
+				tagErrs = append(tagErrs, retrieverError{src: src.ref, err: fmt.Errorf("unable to retrieve source image %s by tag: %v", src.ref, err)})
+				continue
+			}
+			srcDigest := desc.Digest
+			glog.V(3).Infof("Resolved source image %s:%s to %s\n", src.ref, srcTag, srcDigest)
+			src.mergeIntoDigests(srcDigest, pushTargets)
+		}
+
+		canonicalFrom := srcRepo.Named()
+
+		for srcDigestString, pushTargets := range src.digests {
+			// load the manifest
+			srcDigest := digest.Digest(srcDigestString)
+			// var contentDigest digest.Digest / client.ReturnContentDigest(&contentDigest),
+			srcManifest, err := manifests.Get(ctx, digest.Digest(srcDigest), schema2ManifestOnly)
```

Comment:
Why are schema2 manifests excluded?

Reply:
Schema1?  Just because it's older and I was hesitant to bite that off here.  I think it's reasonable to do the conversion during mirroring in the future.